Add additional defensive code within vkUpdateDescriptorSets().

billhollings opened the pull request:
- ~`MVKDescriptorSet::write()`: Revert to verifying each descriptor has the same type as its descriptor binding, to guard against an app mistakenly running off the end of an array of descriptors.~
- ~`MVKDescriptorSet::hasMetalArgumentBuffer()` ensure _layout is available.~

- MVKDescriptorSet::write() & read(): Refactor descriptor iteration to clarify loop indexing, and to check array limits to guard against an app mistakenly running off the end of an array of descriptors.

cdavis5e reviewed (changes requested):
This doesn't seem to have actually helped much with #2283... I'm worried that this could negatively affect the performance of MoltenVK, especially with large descriptor sets. If the bug is in the client, this may not even be necessary.

billhollings commented:
> This doesn't seem to have actually helped much with #2283... I'm worried that this could negatively affect the performance of MoltenVK, especially with large descriptor sets. If the bug is in the client, this may not even be necessary.

Thanks for pointing that out. Yeah, I agree. I've removed the existing code from here.

I do think there is still some validity in guarding against an app mistakenly running off the end of a descriptor binding array, and into descriptors of the wrong type, during a descriptor set update. So I've modified both write and read to check array limits before looping, which will accomplish the same thing the descriptor type checking, without the cost of doing it on each array element.

In doing so, I also refactored `write()` and `read()` to clarify the logic, and the various indexes being iterated.

billhollings commented:
> I do think there is still some validity in guarding against an app mistakenly running off the end of a descriptor binding array, and into descriptors of the wrong type, during a descriptor set update. So I've modified both write and read to check array limits before looping, which will accomplish the same thing the descriptor type checking, without the cost of doing it on each array element.

Oops. This PR went too far. Vulkan requires the ability to write and copy beyond the size of a descriptor binding and into subsequent bindings of the same type, including for inline buffer bindings. PR #2303 reverts this and adds the same capability for inline buffer bindings.
